### ISSUE TITLE
Too many devices mounted on XperiaXZ2c@SFOS4.0.1

### ISSUE THREAD

**Issue by Olf0** [Olf0/mount-sdcard#41]
### Environment
- Device: Xperia XZ2 Compact ("*8314*"?)
- SFOS version: "*clean flash 4.0.1.48, no modifications.*"

### Issue description
"*Storage is mounted normally until installing mount-sdcard, then the issue apears.*" from this [OpenRepos thread](https://openrepos.net/comment/37665#comment-37665), which refers back to this [FSO thread](https://forum.sailfishos.org/t/strange-storage-issue/6252).

### Open question
Which version of *mount-sdcard* was used (might have been v1.6.1 and / or v1.7.2)? 

### Steps for a first analysis
Please execute the following commands:
`devel-su`  # become root
`ls -l /dev/mmcblk[1-9]* /dev/sd[a-z]* /dev/sr* | tee ls-l_dev.txt`
Then upload the resulting file `ls-l_dev.txt` here, or copy & paste its whole content here.

**Comment by levone1** [comment]
open question answer: yes, both versions.

**Comment by levone1** [comment]
output:

brw-rw----    1 root     disk        8,   0 May 29 14:24 /dev/sda
brw-rw----    1 root     disk        8,   1 May 29 14:25 /dev/sda1
brw-rw----    1 root     disk        8,  10 May 29 14:24 /dev/sda10
brw-rw----    1 root     disk        8,  11 May 29 14:24 /dev/sda11
brw-rw----    1 root     disk        8,  12 May 29 14:24 /dev/sda12
brw-rw----    1 root     disk        8,  13 May 29 14:24 /dev/sda13
brw-rw----    1 root     disk        8,  14 May 29 14:24 /dev/sda14
brw-rw----    1 root     disk        8,  15 May 29 14:24 /dev/sda15
brw-rw----    1 root     disk      259,   0 May 29 14:24 /dev/sda16
brw-rw----    1 root     disk      259,   1 May 29 14:24 /dev/sda17
brw-rw----    1 root     disk      259,   2 May 29 14:24 /dev/sda18
brw-rw----    1 root     disk      259,   3 May 29 14:24 /dev/sda19
brw-rw----    1 root     root        8,   2 May 29 14:24 /dev/sda2
brw-rw----    1 root     disk      259,   4 May 29 14:24 /dev/sda20
brw-rw----    1 root     disk      259,   5 May 29 14:24 /dev/sda21
brw-rw----    1 root     disk      259,   6 May 29 14:24 /dev/sda22
brw-rw----    1 root     disk      259,   7 May 29 14:24 /dev/sda23
brw-rw----    1 root     disk      259,   8 May 29 14:24 /dev/sda24
brw-rw----    1 root     disk      259,   9 May 29 14:24 /dev/sda25
brw-rw----    1 root     disk      259,  10 May 29 14:24 /dev/sda26
brw-rw----    1 root     disk      259,  11 May 29 14:24 /dev/sda27
brw-rw----    1 root     disk      259,  12 May 29 14:24 /dev/sda28
brw-rw----    1 root     disk      259,  13 May 29 14:24 /dev/sda29
brw-rw----    1 root     disk        8,   3 May 29 14:24 /dev/sda3
brw-rw----    1 root     disk      259,  14 May 29 14:24 /dev/sda30
brw-rw----    1 root     disk      259,  15 May 29 14:24 /dev/sda31
brw-rw----    1 root     disk      259,  16 May 29 14:24 /dev/sda32
brw-rw----    1 root     root      259,  17 May 29 14:24 /dev/sda33
brw-rw----    1 root     root      259,  18 May 29 14:24 /dev/sda34
brw-rw----    1 root     root      259,  19 May 29 14:24 /dev/sda35
brw-rw----    1 root     disk      259,  20 May 29 14:24 /dev/sda36
brw-rw----    1 root     disk      259,  21 May 29 14:24 /dev/sda37
brw-rw----    1 root     disk      259,  22 May 29 14:24 /dev/sda38
brw-rw----    1 root     disk      259,  23 May 29 14:24 /dev/sda39
brw-rw----    1 root     disk        8,   4 May 29 14:24 /dev/sda4
brw-rw----    1 root     root      259,  24 May 29 14:24 /dev/sda40
brw-rw----    1 root     root      259,  25 May 29 14:24 /dev/sda41
brw-rw----    1 root     root      259,  26 May 29 14:24 /dev/sda42
brw-rw----    1 root     root      259,  27 May 29 14:24 /dev/sda43
brw-rw----    1 root     root      259,  28 May 29 14:24 /dev/sda44
brw-rw----    1 root     root      259,  29 May 29 14:24 /dev/sda45
brw-rw----    1 root     disk      259,  30 May 29 14:24 /dev/sda46
brw-rw----    1 root     disk      259,  31 May 29 14:24 /dev/sda47
brw-rw----    1 root     disk      259,  32 May 29 14:24 /dev/sda48
brw-rw----    1 root     disk      259,  33 May 29 14:24 /dev/sda49
brw-rw----    1 root     root        8,   5 May 29 14:24 /dev/sda5
brw-rw----    1 root     root      259,  34 May 29 14:24 /dev/sda50
brw-rw----    1 root     root      259,  35 May 29 14:24 /dev/sda51
brw-rw----    1 root     root      259,  36 May 29 14:24 /dev/sda52
brw-rw----    1 root     disk      259,  37 May 29 14:24 /dev/sda53
brw-rw----    1 root     disk      259,  38 May 29 14:24 /dev/sda54
brw-rw----    1 root     disk      259,  39 May 29 14:24 /dev/sda55
brw-rw----    1 root     disk      259,  40 May 29 14:24 /dev/sda56
brw-rw----    1 root     disk      259,  41 May 29 14:24 /dev/sda57
brw-rw----    1 root     disk      259,  42 May 29 14:24 /dev/sda58
brw-rw----    1 root     disk      259,  43 May 29 14:24 /dev/sda59
brw-rw----    1 root     disk        8,   6 May 29 14:24 /dev/sda6
brw-rw----    1 root     disk      259,  44 May 29 14:24 /dev/sda60
brw-rw----    1 root     disk      259,  45 May 29 14:24 /dev/sda61
brw-rw----    1 root     disk      259,  46 May 29 14:24 /dev/sda62
brw-rw----    1 root     disk      259,  47 May 29 14:24 /dev/sda63
brw-rw----    1 root     disk      259,  48 May 29 14:24 /dev/sda64
brw-rw----    1 root     disk      259,  49 May 29 14:24 /dev/sda65
brw-rw----    1 root     disk      259,  50 May 29 14:24 /dev/sda66
brw-rw----    1 root     disk      259,  51 May 29 14:24 /dev/sda67
brw-rw----    1 root     root      259,  52 May 29 14:24 /dev/sda68
brw-rw----    1 root     root      259,  53 May 29 14:24 /dev/sda69
brw-rw----    1 root     disk        8,   7 May 29 14:24 /dev/sda7
brw-rw----    1 root     root      259,  54 May 29 14:24 /dev/sda70
brw-rw----    1 root     disk      259,  55 May 29 14:24 /dev/sda71
brw-rw----    1 root     disk      259,  56 May 29 14:24 /dev/sda72
brw-rw----    1 root     disk      259,  57 May 29 14:24 /dev/sda73
brw-rw----    1 root     disk      259,  58 May 29 14:24 /dev/sda74
brw-rw----    1 root     disk        8,   8 May 29 14:24 /dev/sda8
brw-rw----    1 root     disk        8,   9 May 29 14:24 /dev/sda9
brw-rw----    1 root     disk        8,  16 May 29 14:24 /dev/sdb
brw-rw----    1 root     disk        8,  17 May 29 14:24 /dev/sdb1
brw-rw----    1 root     disk        8,  18 May 29 14:24 /dev/sdb2
brw-rw----    1 root     disk        8,  32 May 29 14:24 /dev/sdc
brw-rw----    1 root     disk        8,  33 May 29 14:24 /dev/sdc1
brw-rw----    1 root     disk        8,  34 May 29 14:24 /dev/sdc2

**Comment by Olf0** [comment]
Your command line output shows that the mapping of partition on internal eMMC on this community port of SailfishOS is quite different to all officially supported devices.
* As a quick & dirty workaround, you can manually edit [line 14 in `/etc/udev/rules.d/91-mountsd.rules`](https://github.com/Olf0/mount-sdcard/blob/master/udev/rules.d/91-mountsd.rules#L14) when you have *mount-sdcard* installed: Alter <br />it to `KERNEL!="mmcblk*|sd[d-z]*|sr*", GOTO="mountsd_end"` <br />(from `KERNEL!="mmcblk[1-9]*|sd*|sr*", GOTO="mountsd_end"`)
* For a proper fix, I think it is wise to consult the maintainer of the device adaptation (i.e., "SailfishOS porter") you are using, first.<br />
  This is Rinigus' adaptation for the XperiaXZ2c you are using, right?<br />
  I.e., https://github.com/sailfishos-sony-tama / https://build.merproject.org/project/show/nemo:devel:hw:sony:tama ?

**Comment by Olf0** [comment]
Side note & off topic ("OT"):
> Also, one thing I was interested in installing it for is the info that says it will enable sd card access on Android, but not working.

Supposedly you are addressing the statement "*Ensure, that AlienDalvik begins starting after mounting succeeded, to allow for [android_storage on SD-card](https://together.jolla.com/question/203539/guide-externalising-android_storage-and-other-directories-files-to-sd-card/#203539-2-externalising-homenemoandroid_storage).*" (from the third bullet point of https://openrepos.net/content/olf/mount-sdcard).
If so, you heavily misread it ("to allow for" points to a prerequisite) and did not follow its link or did not read the content there (including the comments).
Tl;DR: Currently that does not work with AlienDalvik 8, 9 or 10, because Jolla changes something, and nobody has traced what and how to workaround it, yet.

**Comment by Olf0** [comment]
#### Back on topic, again
Was [your output](https://github.com/Olf0/mount-sdcard/issues/41#issuecomment-850878673) generated with or without an SD-card inserted?

Well, what I really want to know is: To which devices names does (partitions on) an SD-card become assigned?
You might look at the difference of above output with and without an SD-card inserted and report that "delta" here.

**Comment by rinigus** [comment]
On Sony Tama (XZ2{c}/3), system partitions are on /dev/sd{a,b} and maybe few more. SD card is mounted on `mmcblk0`. Same, ass far as I know, goes for Pro1 and OnePlus 5 or 6 (or both?). Issue has been also exposed in systems settings and was "fixed" there 

- https://github.com/sailfishos-sony-tama/nemo-qml-plugin-systemsettings/commit/dd5f1fa4042f0744e0b3b7f1483d884895fe05fc
- https://github.com/sailfishos-sony-tama/nemo-qml-plugin-systemsettings/commit/4ea92c2298fea07224660fe3fed5a64783fe6895

We have been asking few times for getting some proper fix using udev, but nothing happened so far.

**Comment by Olf0** [comment]
@rinigus, O.K., I think I can handle that ("*system partitions are on /dev/sd{a,b}*"), but "*and maybe few more*" might be tricky.

But "*SD card is mounted on mmcblk0*" collides with the system partitions on all officially supported devices, which are *all* at some `mmcblk0p*`!
The crucial point here is the inability to distinguish a internal, fixed eMMC from an SD- or MMC-card inserted into some internal slot, because they all are attached via an MMC/SD-controller (as a udev `SUBSYSTEMS=`).

An which device names does USB-attached storage become assigned to?
If the answer is "on some /dev/sd?, after the last *system partitions [...] on /dev/sd{a,b} and maybe few more*", this scheme is very awkward and hard to utilise (but seems feasible here, in contrast to `mmcblk0`, because one can test for `SUBSYSTEMS="usb"`).

**Comment by Olf0** [comment]
After thinking about it a little, I believe this device naming scheme has the potential to cause big trouble (up to bricking devices non-manually; *mount-sdcard* currently just enables one to do that manually), because it more or less conflicts with what Jolla chose for their device adaptations.
Insofar I am glad that @levone1 discovered this.

Hence I will consider to limit the `/dev/sd[a-z]*` *mount-sdcard* and *crypto-sdcard* care about to those with `SUBSYSTEMS="usb"`, but I think the proper fix is to move those community device adaptations to Jolla's device naming scheme.

**Comment by rinigus** [comment]
The naming scheme for devices comes from Android layer/SOC and cannot be changed. The devices in question are using SD 835, 845 and I suspect that the later SDs as well (don't know though). So, in my eyes, real solution would be to instruct UDEV to attach some label to removable devices and use that label to handle mounting.

I agree, it is a dangerous situation and users should be aware of it.

**Comment by rinigus** [comment]
PS: I have not checked specific devices as I didn't have time for it. But, in principle, we should not look for solution where each app or subsystem has to write down all possible configs. Hence I wasn't documenting it here.

**Comment by Olf0** [comment]
> So, in my eyes, real solution would be to instruct UDEV to attach some label to removable devices and use that label to handle mounting.

Oh, that would be the perfect solution.  But only Jolla and the udev project are in the position to change that for SailfishOS (or all distros using udev) in general.  Mmmh, the pathn via the udev project might be the quicker one.

> But, in principle, we should not look for solution where each app or subsystem has to write down all possible configs.

I fully agree, but what else can be done to fix that soon?
Which implies "on our own behalf" (your, mine or together; i.e. without having to involve others).

**Comment by Olf0** [comment]
O.K., I drafted a workaround per https://github.com/Olf0/mount-sdcard/commit/89bc3d1f921f8ec858bfb3a67907c6986fea3f65
I will try to release a new version within the next 10 days, which will be named *mount-sdcard 1.7.3*.  @levone1, I would be glad, if you test it then, for that I will explicitly notify you to the Github release, here (ususally one to two days before I release at Openrepos).

But as I wrote
> But "SD card is mounted on mmcblk0" collides with the system partitions on all officially supported devices, which are all at some mmcblk0p*!
The crucial point here is the inability to distinguish a internal, fixed eMMC from an SD- or MMC-card inserted into some internal slot, because they all are attached via an MMC/SD-controller

@rinigus, is it possible to assign the sd-card slot to mmcblk1, or to put it more generic: Can you block mmcblk**0** from being used for the device enumeration on your SFOS port(s)?
Side note: Mind that SD- and MMC-cards in USB-attached card readers also become assigned to mmcblk* names.

**Comment by rinigus** [comment]
I am massively swamped right now and will probably be next week as well. After that I can have a meaningful discussion on it. Sorry for it.

Don't count on us to be able to block mmcblk0 from appearing. However, maybe we could tag it somehow as removable device. I will have to pull my notes regarding it, hopefully I wrote it down in the corresponding issue of Tama port. But that will have to wait.

**Comment by Olf0** [comment]
> Sorry for it.

**No!** ;-)
That is and has to be always O.K., especially for everybody spending his / her spare time on this.
Just keep in mind that the few handful of Jolla employees in technical jobs have no problem to delay various things indefinitely, leaving unfinished task in the state they are forever etc. etc. without ever saying a word: They are getting paid for that work, still you will never ever hear an excuse, or even an apology from them.

TL;DR: Ping us here, when you have some time to deal with this.

Side note: If "we" (i.e., you for your ports) are able to "tag" (should not be a real udev *tag*, rather just a udev *variable*), I can (and am willing to) evaluate that, but then we are back at the very individual solution you initially wanted to avoid (IMO, correctly so).

**Comment by Olf0** [comment]
@levone1, please test [mount-sdcard-1.8.0-1.sfos340.noarch.rpm](https://github.com/Olf0/mount-sdcard/releases/download/1.8.0-1.sfos340/mount-sdcard-1.8.0-1.sfos340.noarch.rpm)

Suggested procedure:
Start at the command line as

**Comment by Olf0** [comment]
### [OT] continued
@levone1 , as Jolla has been fixing stuff lately and had explicitly broken SD-card access for the AlienDalvik runtime environment in SailfishOS 3.3.0 - 3.4.0, you may just retry the suggestions in this sub-thread (please read it twice in its completeness, which is just a few messages, *before* starting to test / try things): https://together.jolla.com/question/203539/guide-externalising-android_storage-and-other-directories-files-to-sd-card/?answer=206350#post-id-206350

**Comment by Olf0** [comment]
@levone1, as you were so keen on having this quickly addressed and fixed:
Can you [please test and report back](https://github.com/Olf0/mount-sdcard/issues/41#issuecomment-851626177) here!

**Comment by Olf0** [comment]
Closing, as there is no feedback by the original reporter and because this is likely working on this niche devices now.
A bit unfortunate to receive no feedback at all after all this upheaval in many places (FSO, TMO, OpenRepos and here), but this did not come completely unexpected (rather somewhat expectedly).

@rinigus, please utilise this (closed) thread to notify me, when you have found a way to somehow mark a "removable" `mmcblk0*` as such.

**Comment by levone1** [comment]
Update installed and working fine - no additional "removable media" devices. Thanks

**Comment by Olf0** [comment]
@levone1, thanks for the feedback.

As you asked, two things were problematic for me:
- [Technical] If it comes to my knowledge that a tool I provide enables users to wreck / "brick" their devices, I feel some responsibility to fully comprehend the issue and (if necessary, feasible, able to) do something about it.  Despite being FLOSS licensed (i.e., no liability at all).
- [Procedural] If you would have filed an issue report here, let me (or in general: the maintainer) handle your report, simply answered the question(s) and ultimately test a new release, this would have been far less tedious for me.
  It is O.K., that you might not see a couple of non-technical aspects your "spreading the news of something going pretty wrong in many places" has, but that is definitely not helpful!  There is usually a single best place for this: The primary issue tracker of the original project / its maintainer.

**Comment by Olf0** [comment]
P.S.: And as stated, @levone1, if you like to try out things / contribute to a long lost open question / serve your original interest of having `android_storage` on SD-card, take a look at the two "Off Topic (OT)" postings above.

**Comment by levone1** [comment]
Externalizing seems interesting ... I'll probably mess with it at some point to test. My current sd card is nearly full, but I'll pull one out to work with.  Are yo saying that as of now, there is no possible way for AD to access the card directly?

**Comment by Olf0** [comment]
???
Please try to precisely describe what you mean, especially because we all are (presumably) non-native English speakers and writers:
1. > Externalizing seems interesting ...
   
   Does that address "externalising `android_storage` to SD-card"?
2. > Are yo saying that as of now, there is no possible way for AD to access the card directly?
   
   I don't know, because I fail to comprehend, what you mean with "directly".
   I wrote "[*Jolla [...] had explicitly broken SD-card access for the AlienDalvik runtime environment in SailfishOS 3.3.0 - 3.4.0, [...]*](https://github.com/Olf0/mount-sdcard/issues/41#issuecomment-851667802)", hence "externalising `android_storage` to SD-card" cannot work on SailfishOS 3.3.0 - 3.4.0.
  This is all I know, any further research (i.e. trying out and analysing things, reading documentation, etc.: in short, "research") is up to you.

**Comment by levone1** [comment]
My initial fiddling ...
I went through section 2 (Externalising /home/nemo/android_storage) of your guide, and seemed to be working properly, (replacing 'nemo' with 'defaultuser'), up to section 2.2, (Integrating an externalised android_storage when booting), where I dead-ended, since there is no longer such a thing as /opt/alien/system/script, (the only thing in .../alien is system.img). I did unsquash syste img to poke around, thinking maybe there was some correspondent script in there somewhere, but didn't find anything.

Then I tested the guide by mynameisnotimportant posted under yours, and I found a significant difference in my config file compared to his. I tried the tweaks he posted anyway, just to see, but it caused AD to not start, (tried a few different paths/syntaxes).

Here is the content of my AD config file:
`lxc.rootfs.path = loop:/opt/alien/system.img
lxc.uts.name = aliendalvik
lxc.arch = aarch64

lxc.init.cmd = /init

lxc.mount.auto = cgroup:ro sys:ro

# necessary dev nodes (renamed)
lxc.mount.entry = /dev/puddlejumper dev/binder none bind,create=file 0 0
lxc.mount.entry = /dev/vndpuddlejumper dev/vndbinder none bind,create=file 0 0

# necessary device nodes (same name)
lxc.mount.entry = /dev/ashmem dev/ashmem none bind,create=file 0 0

lxc.mount.entry = /dev/fuse dev/fuse none bind,create=file 0 0
lxc.mount.entry = /dev/ion dev/ion none bind,create=file,optional 0 0

# Mount /data - always /home/.android now
lxc.mount.entry = tmpfs mnt tmpfs mode=0755,uid=0,gid=1000

# useful device properties from the host
lxc.mount.entry = /tmp/aliendalvik/generated_props system/vendor/alien.prop none bind,create=file 0 0

# Xkb - required for on screen keyboard
lxc.mount.entry = /usr/share/X11/xkb usr/share/X11/xkb none bind 0 0

lxc.include = /var/lib/lxc/aliendalvik/bsp_config
lxc.include = /var/lib/lxc/aliendalvik/extra_config
  
`
It's true that 'accessing' sd card from Android works, but read-only, (at least for me)...

**Comment by Olf0** [comment]
> My initial fiddling ...

... is still off-topic to this issue thread: Please open a new one.
Please do choose the title and issue description well, so that any new reader is able to comprehend it easily.  I.e., provide context or well labelled links to the necessary context.

> I went through section 2 (Externalising /home/nemo/android_storage) of your guide, and seemed to be working properly, (replacing 'nemo' with 'defaultuser'), up to section 2.2, (Integrating an externalised android_storage when booting), where I dead-ended, since there is no longer such a thing as /opt/alien/system/script, (the only thing in .../alien is system.img).

Well, my advice is still to wholly read threads, before starting to experiment / trying to reproduce.
As clearly stated, this guide was created for 1st and 2nd generation AlienDalvik (the ASOP 4 based ones), only.

> I did unsquash syste img to poke around, thinking maybe there was some correspondent script in there somewhere, but didn't find anything.

Structurally I expected that outcome, but thanks for checking.

> Then I tested the guide by mynameisnotimportant posted under yours, 

... where I pointed you to, twice.
 
> and I found a significant difference in my config file compared to his.

Which ones?

> I tried the tweaks he posted anyway, just to see, but it caused AD to not start, (tried a few different paths/syntaxes).
> 
> Here is the content of my AD config file:

The original or patched content?

Also, please upload larger (> ~300 Bytes) files as files attachments to a post.
If you post file content inline, protect it from being interpreted as markdown by quoting.

> It's true that 'accessing' sd card from Android works, but read-only, (at least for me)...

Thanks for this result.
Yes, that makes sense to me, when I look at Jolla's AlienDalvik-LXC configuration.
Which is unfortunate, because it once was read-write.
And this has to be resolved first, before trying to put `android_storage` onto SD-card.
